Update MAVLink parsing for wiki inclusion

ES-Alexander opened the pull request:
ArduPilot Wiki uses reStructuredText (rst) source files, so supporting rst output is essential for this script to generate wiki files. Ideally this can be run automatically for every vehicle type and version, like the existing parameter lists.

- [x] Add rst support
- [ ] Add per-vehicle dropdown version selection support, like the parameter lists
    - Not yet sure how this is done
- [ ] Set up automatic generation per vehicle, per version
    - May be done via autotest, [like for parameters](https://github.com/ArduPilot/ardupilot/blob/master/Tools/autotest/autotest.py#L194)
        - Moving this to autotest could be useful for notifying of changes in MAVLink support on PRs, but is probably too much effort for this PR
    - Will likely get done externally, in what is currently [the parameters repository](https://github.com/ArduPilot/ParameterRepository/blob/main/scripts/run_param_parse.py#L132)
        - Extra attractive because the script requires Python &geq; 3.11, which will be easier to specify there than forcing autotest to update

Example file for ArduSub 4.1, with all options included:
[ArduSub_Sub-4.1_MAVLink_Messages.rst.zip](https://github.com/user-attachments/files/15531781/ArduSub_Sub-4.1_MAVLink_Messages.rst.zip)

amilcarlucas commented:
This seams related to this file that I have been working on for the past weeks:

```python
#!/usr/bin/python3

'''
Rsync apm.pdef.xml files for different versions of the ArduPilot firmware
For each version, it checks out the corresponding tag, generates parameter metadata,
and finally rsyncs the updated parameter metadata pdef.xml files.

(C) 2024 Amilcar do Carmo Lucas

SPDX-License-Identifier:    GPL-3
'''

import os
import getpass
import datetime
import shutil
import subprocess

VEHICLE_TYPES = ["Copter", "Plane", "Rover", "ArduSub", "Tracker"]  # Add future vehicle types here
RSYNC_USERNAME = 'amilcar'

# Store the current working directory
old_cwd = os.getcwd()

def get_vehicle_tags(vehicle_type):
    """
    Lists all tags in the ArduPilot repository that start with the given vehicle type followed by '-'.
    Returns a list of tag names.
    """
    try:
        # Change to the ArduPilot directory
        os.chdir('../ardupilot/')
        tags_output = subprocess.check_output(['git', 'tag', '--list', f'{vehicle_type}-[0-9]\\.[0-9]\\.[0-9]'], text=True)
        return tags_output.splitlines()
    except Exception as e: # pylint: disable=broad-exception-caught
        print(f"Error getting {vehicle_type} tags: {e}")
        return []

def generate_vehicle_versions():
    """
    Generates arrays for each vehicle type based on the tags fetched from the ArduPilot repository.
    """
    vehicle_versions = {}

    for vehicle_type in VEHICLE_TYPES:
        tags = get_vehicle_tags(vehicle_type)
        if tags:
            vehicle_versions[vehicle_type] = [tag.split('-')[1] for tag in tags]

    return vehicle_versions

def create_one_pdef_xml_file(vehicle_type: str, dst_dir: str, git_tag: str):
    os.chdir('../ardupilot')
    subprocess.run(['git', 'checkout', git_tag], check=True)
    # subprocess.run(['git', 'pull'], check=True)
    subprocess.run(['Tools/autotest/param_metadata/param_parse.py', '--vehicle', vehicle_type, '--format', 'xml'], check=True)
    # Return to the old working directory
    os.chdir(old_cwd)

    if not os.path.exists(dst_dir):
        os.makedirs(dst_dir)

    # Insert an XML comment on line 3 in the ../ardupilot/apm.pdef.xml file to indicate
    # the tag used to generate the file and the current date
    with open('../ardupilot/apm.pdef.xml', 'r', encoding='utf-8') as f:
        lines = f.readlines()
    lines.insert(2, f'<!-- Generated from git tag {git_tag} on {datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")} -->\n')
    with open('../ardupilot/apm.pdef.xml', 'w', encoding='utf-8') as f:
        f.writelines(lines)
    shutil.copy('../ardupilot/apm.pdef.xml', f'{dst_dir}/apm.pdef.xml')

# Function to sync files using rsync
def sync_to_remote(vehicle_type: str) -> None:
    src_dir = f'{vehicle_type}/'
    dst_user = RSYNC_USERNAME
    dst_host = 'firmware.ardupilot.org'
    dst_path = f'param_versioned/{vehicle_type}/'

    # Construct the rsync command
    rsync_cmd = [
        'rsync',
        '-avz',
        '--progress',
        '--password-file=-',  # Use stdin for password
        src_dir,
        f'{dst_user}@{dst_host}::{dst_path}'
    ]

    print(f'Synchronizing {src_dir} to {dst_path}...')
    print(rsync_cmd)
    subprocess.run(rsync_cmd, check=True)

def main():
    vehicle_versions = generate_vehicle_versions()

    # Prompt for the password securely
    _remote_password = getpass.getpass("Enter the remote server password: ")

    # Iterate over the vehicle_versions list
    for vehicle_type, versions in vehicle_versions.items():
        for version in versions:

            if version[0] == '3' and vehicle_type != 'AP_Periph':
                continue # Skip ArduPilot 3.x versions, as param_parse.py does not support them out of the box
            if version[0] == '4' and version[2] == '0' and vehicle_type != 'ArduSub':
                continue # Skip ArduPilot 4.0.x versions, as param_parse.py does not support them out of the box
            vehicle_dir = vehicle_type

            if vehicle_type == 'ArduSub':
                vehicle_dir = 'Sub'

            create_one_pdef_xml_file(vehicle_type, f'{vehicle_dir}/stable-{version}', f'{vehicle_type}-{version}')
        sync_to_remote(vehicle_type)

if __name__ == '__main__':
    main()
```

Maybe we can combine this together?

ES-Alexander commented:
@amilcarlucas where/when are you expecting/wanting yours to run?

As far as I'm aware the [parameters repository](https://github.com/ArduPilot/ParameterRepository/tree/main) already generates all the `apm.pdef.xml` files (with CI checks for new tags to process every 5 days), so I expect it could be reasonably straightforward to add your rsync updates [there](https://github.com/ArduPilot/ParameterRepository/blob/main/scripts/run_param_parse.py#L145) as just an extra step?

I'm not sure how much sense it would make to combine efforts on a PR in that repo, given your feature is an additional endpoint for the parameters, whereas mine is a new script to run and store the outputs of, but if you're also wanting to include the MAVLink message support in your rsync transfers then I suppose that ties them together somewhat?

amilcarlucas commented:
Nevermind, I jumped to conclusions.

ES-Alexander commented:
Did some more digging and found that adding a version selection dropdown makes more sense to handle as a patch during generation in the specific applications where it's relevant. Since both of the remaining "planned" components of this PR can be handled in separate repositories, I think this is ready for review / can be merged in 🙂

ES-Alexander commented:
Not really sure who should review this PR, but @Hwurzburg might have some ideas or preferences around the formatting of the generated files 🤷‍♂️

Hwurzburg commented:
Was all the formatting in the example sub file generated by this script?

ES-Alexander commented:
> Was all the formatting in the example sub file generated by this script?

Yeah - the `.rst` example file is just directly from the script output :-)

I generated it from the `Sub-4.1` branch with the following invocation:
```
python mavlink_parse.py -v ArduSub --format rst -b Sub-4.1 --header ardupilot_wiki -gucq
```

(noting once again that the script requires Python 3.11 to run)

Hwurzburg commented:
I've tagged @peterbarker and @IamPete1  to review if the groupings and generated data is correct....with PeterB's help I could add per vehicle pages into the Dev sites wiki under a MAVLink commands topic....

ES-Alexander commented:
> I've tagged @peterbarker and @IamPete1 to review if the groupings and generated data is correct

I've just rebased over master, and generated some more rst files for different vehicle types, to make that checking a bit more straightforward:

[MultiVehicle_master_MAVLink_Messages.zip](https://github.com/user-attachments/files/15793798/MultiVehicle_master_MAVLink_Messages.zip)

> with PeterB's help I could add per vehicle pages into the Dev sites wiki under a MAVLink commands topic....

Nice! 😃 
I'm happy to make a PR to generate these automatically in what is currently the parameters repository, if that's helpful, but also happy for some other approach to be taken.

I'd note that because the MAVLink protocol updates, and I'm currently including the "unsupported messages" in the output generation, the outputs for each version should ideally be re-generated each time with the latest script version, and the most recent pymavlink release, rather than tying the script to the vehicle branch and only generating once per version. That said, that's not a critical requirement, and generating once per version could be ok if we either accept that newer messages won't show up, or just don't generate the unsupported messages - I just like the idea of being able to search for a message and finding out clearly if it is definitely unsupported 🤷‍♂️

Hwurzburg commented:
having to manually update the page generation once in a while is fine for now, IMO....a PR to generate the page would be good, then I can add it to the Dev wiki.....very few users, and only devs, will be using it but it will certainly help them, even if occasionally out of date...

Hwurzburg commented:
sorry its taken me so long to get back to this, but I wanted to add wiki files for Plane,Copter,Rover messages as of current master....the params repo does not even have files for 4.5...could you generate these three and I will ad to dev wiki...thanks

ES-Alexander commented:
> sorry its taken me so long to get back to this

I get it - unfortunately the list of interesting/useful things to do seems to grow much faster than the available time to do things 😅

> could you generate these three and I will ad to dev wiki

It's been a while since I did those initial generations, so I'll need to figure out the relevant setup/process again. I'm hoping it's mostly covered by the Action I made in the parameters repo, although from memory that's reliant on this PR being merged, so I might need to make some custom branches to do the manual generations for now.

I'll try to sort that out this week :-)

> I wanted to add wiki files for Plane,Copter,Rover messages as of current master....the params repo does not even have files for 4.5...could you generate these three

To confirm, did you want files for just the 4.5 release, or current master as well?
If I get to doing this before you respond I'll just generate both - I don't expect it to require much extra effort once I have the process working :-)

Hwurzburg commented:
4.6 is about to drop...so master only will be fine

ES-Alexander commented:
Just got around to this.

> 4.6 is about to drop...so master only will be fine

Here ya go 🙂
[MultiVehicle_master_MAVLink_Messages_Nov-5.zip](https://github.com/user-attachments/files/17621527/MultiVehicle_master_MAVLink_Messages_Nov-5.zip) 

> the params repo does not even have files for 4.5

Following up on this, it _did_ have 4.5 files, [per my initial pull request to that repo](https://github.com/ArduPilot/ParameterRepository/pull/7/files). Those files then got auto-removed as part of a recent metadata update, and weren't regenerated because [the process for doing so](https://github.com/ArduPilot/ParameterRepository/pull/8) is reliant on this PR being merged so the relevant parsing script option (rst format) is available. Until this PR is merged that will continue to happen, unless I make a hacky workaround that references a branch in my ardupilot fork, which would then need to be updated again once this PR is merged.

Given you're willing to put its outputs in the wiki, is there anything holding back this PR from getting merged?

Hwurzburg commented:
@ES-Alexander PR for wiki inclusion...thanks https://github.com/ArduPilot/ardupilot_wiki/pull/6379

ES-Alexander commented:
> PR for wiki inclusion...[ArduPilot/ardupilot_wiki#6379](https://github.com/ArduPilot/ardupilot_wiki/pull/6379)

Nice! Thanks for adding those in :-)

> ... had to reformat a few things in the files.

The generated files? If so, do you remember what you had to change, and are they things we should be able to edit the generator script to provide from the start?

Hwurzburg commented:
the tag must be different in each file....I retitled the headers

ES-Alexander commented:
I have a vague recollection I made those the same on purpose, with the idea they'd be replaced automatically as part of the version selection support. That said, I may be misremembering, and it should be possible to at least roughly do something similar as part of the initial generator script if we want to, since it does already know the vehicle(s) being generated for, and the branch.

Hwurzburg commented:
not required at this point, I can manually run every major release to keep up to date.....going to merge as is...thanks

tridge commented on Tools/scripts/mavlink_parse.py:
```diff
@@ -7,7 +7,7 @@
 
 from pymavlink.dialects.v20 import (
-    common, icarous, cubepilot, uAvionix, ardupilotmega
+    common, icarous, cubepilot, uAvionix, ardupilotmega, development
```
should be just be 'all'? either that or needs minimal
does this end up with lots of duplicates via the recursion of ardupilot -> common -> minimal?

ES-Alexander replied:
> should be just be 'all'?

It's nice to break them out into dialects, so that the message definitions can be linked to directly within the MAVLink documentation.

> either that or needs minimal

The messages defined in `minimal` [are duplicated in the `common` doc](https://mavlink.io/en/messages/common.html#HEARTBEAT), so I didn't think it was necessary to break it out further (because the links already work fine). 

Happy to make that change if you think it's preferable to link to the minimal docs instead of the common docs for those couple of messages. Thoughts?

> does this end up with lots of duplicates via the recursion of ardupilot -> common -> minimal?

Nope - the dialects are checked in order (see `determine_dialect`), so it just stops when it gets to the first one that matches. The `MAVLinkDetector` class maps relevant messages to their dialects (and corresponding documentation links) as needed, so the only message duplicates are if the firmware handles the same message in different ways that are checked for (e.g. `SYSTEM_TIME` can be incoming, (automatically) outgoing, and also requestable, so appears in all three tables).

ES-Alexander replied:
I've refactored the import code to be dynamic, based on the enum, and added in `minimal` (and the other currently missing message sets) so it shows up correctly.

tridge reviewed (changes requested):
needs to handle recursion

peterbarker commented on Tools/scripts/mavlink_parse.py:
```diff
@@ -193,7 +198,12 @@ def __init__(self, common_files, vehicle='ALL',
+
+            try:
+                text = file.read_text()
+            except FileNotFoundError:  # Broken symlink
+                continue
```
Why would we ignore this?!

peterbarker commented on Tools/scripts/mavlink_parse.py:
```diff
@@ -7,7 +7,7 @@
 
 from pymavlink.dialects.v20 import (
-    common, icarous, cubepilot, uAvionix, ardupilotmega
+    common, icarous, cubepilot, uAvionix, ardupilotmega, development
```
Missing minimal.xml and/or parsing of include files and use all.xml?

Check Wiki to see if HEARTBEAT is missing?

peterbarker reviewed (changes requested):
maybe just try `all.xml`

ES-Alexander commented:
To anyone following along, I've just:
- Rebased this over master
- Cleaned up some CLI argument descriptions
- Added ID numbers to the first column of the tables
    - That was previously [added independently in the docs](https://github.com/ArduPilot/ardupilot_wiki/pull/6789), so it seemed to make sense to include in future generations

peterbarker reviewed (approved):
LGTM.

@Hwurzburg  has to be happy, 'though.
